### DISCUSSION TITLE
OfType for polymorphic collections

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#15764] [Q&A] [answered]
Is it possible to narrow down polymorphic collections with an OfType<> filter or similar?

Taking the example 
https://ravendb.net/docs/article-page/5.3/csharp/indexes/indexing-polymorphic-data

and
```
DocumentStore store = new DocumentStore()
{
    Conventions =
    {
        FindCollectionName = type =>
        {
            if (typeof(Animal).IsAssignableFrom(type))
                return "Animals";
            return DocumentConventions.DefaultGetCollectionName(type);
        }
    }
};
``` 

The query
``` 
from cat in DbSession.Query<Cat>()
    where Name = name 
    select cat;
``` 
will return also dogs.

If animal would have an `Owner`, where the owner is 
```
class CatLover : Owner {}
```
or
```
class DogLover : Owner {}
```

I noticed that `OfType<CatLover>()` in the query will serialize to a type check that is sent to the database.
Is something similar possible that will serialize in a metadata type comparison on the query?

**Comment by ayende** [comment, marked as answer]
At the RavenDB Server level, there is no such option, since the data doesn't exists in the server.
We have the `Raven-Clr-Type`, but not the inheritance hierarchy.

Your option may be to add that to the query, so you'll have a property called `Type` or something that would be set on save.

Here is a trivial example:

```
public List<string> InheritanceHierarchy
{
    get
    {
        var hierarchy = new List<string>();
        var type = this.GetType();
        while(type != typeof(object))
        {
            hierarchy.Add(type.Name);
            type = type.BaseType;
        }
        return hierarchy;
    }
}
``` 

And then you can do a query by any of the items in the hierarchy.

**Reply by danielwinkler** [reply]
Thanks for the reply Oren!

I did not intend to query the hierarchy, but just want to narrow the query to the exact type as stored in the metadata (Raven-Clr-Type).
Is there an (elegant) way to include the Raven-Clr-Type match in the Query?

**Reply by ayende** [reply]
Not easily, no. We don't have a way to expose that to the Linq provider, I'm afraid.
You'll need to define a `Type` property in the baseclass, which will assist here.

**Reply by danielwinkler** [reply]
Thanks!
